### ISSUE TITLE
[Bug] 0.12.0版本开启raft，无法成功启动

### ISSUE THREAD

**Issue by liweidong1216** [hugegraph/hugegraph#1726]
### Bug Type (问题类型)

server status (启动/运行异常)

### Before submit

- [X] 我已经确认现有的 [Issues](https://github.com/hugegraph/hugegraph/issues) 与 [FAQ](https://hugegraph.github.io/hugegraph-doc/guides/faq.html) 中没有相同 / 重复问题

### Environment (环境信息)

- Server Version: v0.11.x
- Backend: RocksDB x nodes, HDD or SSD 
- OS: xx CPUs, xx G RAM, Centos 7.x 
- Data Size:  xx vertices, xx edges 

### Expected & Actual behavior (期望与实际表现)

下载了最新的0.12.0版本(近几个版本都出现了这个错误)，开启了raft，无法成功启动，配置如下：
# gremlin entrance to create graph
# auth config: com.baidu.hugegraph.auth.HugeFactoryAuthProxy
gremlin.graph=com.baidu.hugegraph.HugeFactory

# cache config
#schema.cache_capacity=100000
# vertex-cache default is 1000w, 10min expired
vertex.cache_type=l2
#vertex.cache_capacity=10000000
#vertex.cache_expire=600
# edge-cache default is 100w, 10min expired
edge.cache_type=l2
#edge.cache_capacity=1000000
#edge.cache_expire=600

# schema illegal name template
#schema.illegal_name_regex=\s+|~.*

#vertex.default_label=vertex

backend=rocksdb
serializer=binary

store=hugegraph

raft.mode=true
raft.safe_read=false
raft.use_snapshot=false
raft.endpoint=172.24.4.84:8281
raft.group_peers=172.24.4.84:8281,172.24.4.85:8281,172.24.4.86:8281
raft.path=./raft-log
raft.use_replicator_pipeline=true
raft.election_timeout=10000
raft.snapshot_interval=3600
raft.backend_threads=48
raft.read_index_threads=8
raft.read_strategy=ReadOnlyLeaseBased
raft.queue_size=16384
raft.queue_publish_timeout=60
raft.apply_batch=1
raft.rpc_threads=80
raft.rpc_connect_timeout=5000
raft.rpc_timeout=60000

search.text_analyzer=jieba
search.text_analyzer_mode=INDEX

# rocksdb backend config
#rocksdb.data_path=/path/to/disk
#rocksdb.wal_path=/path/to/disk

# cassandra backend config
cassandra.host=localhost
cassandra.port=9042
cassandra.username=
cassandra.password=
#cassandra.connect_timeout=5
#cassandra.read_timeout=20
#cassandra.keyspace.strategy=SimpleStrategy
#cassandra.keyspace.replication=3

# hbase backend config
#hbase.hosts=localhost
#hbase.port=2181
#hbase.znode_parent=/hbase
#hbase.threads_max=64

# mysql backend config
#jdbc.driver=com.mysql.jdbc.Driver
#jdbc.url=jdbc:mysql://127.0.0.1:3306
#jdbc.username=root
#jdbc.password=
#jdbc.reconnect_max_times=3
#jdbc.reconnect_interval=3
#jdbc.sslmode=false

# postgresql & cockroachdb backend config
#jdbc.driver=org.postgresql.Driver
#jdbc.url=jdbc:postgresql://localhost:5432/
#jdbc.username=postgres
#jdbc.password=
#jdbc.postgresql.connect_database=template1

# palo backend config
#palo.host=127.0.0.1
#palo.poll_interval=10
#palo.temp_dir=./palo-data
#palo.file_limit_size=32

可以成功执行init-store.sh
确认lib目录下存在jraft-core-1.3.5.jar
执行start-hugegrpah.sh的时候出现以下错误
2022-01-06 10:49:58 [main] [ERROR] c.b.h.d.HugeGraphServer - HugeRestServer start error: 
java.lang.NoSuchMethodError: org.rocksdb.BloomFilter.<init>(IZ)V
        at com.alipay.sofa.jraft.util.StorageOptionsFactory.getDefaultRocksDBTableConfig(StorageOptionsFactory.java:303) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.util.StorageOptionsFactory.getRocksDBTableFormatConfig(StorageOptionsFactory.java:289) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.storage.impl.RocksDBLogStorage.createColumnFamilyOptions(RocksDBLogStorage.java:173) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.storage.impl.RocksDBLogStorage.initAndLoad(RocksDBLogStorage.java:219) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.storage.impl.RocksDBLogStorage.init(RocksDBLogStorage.java:205) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.storage.impl.RocksDBLogStorage.init(RocksDBLogStorage.java:70) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.storage.impl.LogManagerImpl.init(LogManagerImpl.java:184) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.storage.impl.LogManagerImpl.init(LogManagerImpl.java:77) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.core.NodeImpl.initLogStorage(NodeImpl.java:581) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.core.NodeImpl.init(NodeImpl.java:991) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.core.NodeImpl.init(NodeImpl.java:138) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.RaftServiceFactory.createAndInitRaftNode(RaftServiceFactory.java:47) ~[jraft-core-1.3.5.jar:?]
        at com.alipay.sofa.jraft.RaftGroupService.start(RaftGroupService.java:129) ~[jraft-core-1.3.5.jar:?]
        at com.baidu.hugegraph.backend.store.raft.RaftNode.initRaftNode(RaftNode.java:125) ~[hugegraph-core-0.12.0.jar:0.12.0.0]
        at com.baidu.hugegraph.backend.store.raft.RaftNode.<init>(RaftNode.java:60) ~[hugegraph-core-0.12.0.jar:0.12.0.0]
        at com.baidu.hugegraph.backend.store.raft.RaftSharedContext.initRaftNode(RaftSharedContext.java:138) ~[hugegraph-core-0.12.0.jar:0.12.0.0]
        at com.baidu.hugegraph.backend.store.raft.RaftBackendStoreProvider.waitStoreStarted(RaftBackendStoreProvider.java:142) ~[hugegraph-core-0.12.0.jar:0.12.0.0]

**Comment by imbajin** [comment]
raft 版本还有几个 patch 需要合入, 之后会再更新一个版本. 现在处于 beta, 多谢反馈

**Comment by liweidong1216** [comment]
使用最新的0.12-b1的源码重新编译打包，只要开启了raft，在执行start-hugegraph.sh以后，出现同样的错误java.lang.NoSuchMethodError: org.rocksdb.BloomFilter.(IZ)V

**Comment by imbajin** [comment]
> 使用最新的 0.12-b1 的源码重新编译打包，只要开启了 raft，在执行 start-hugegraph.sh 以后，出现同样的错误 java.lang.NoSuchMethodError: org.rocksdb.BloomFilter.(IZ) V

b1 是什么版本? 可以确认一下 md5 值, 可能下到 old 的版本了?

![image](https://user-images.githubusercontent.com/17706099/149872304-468f07bf-c993-40b7-ae4e-ab6e1e28e2ba.png)

**Comment by liweidong1216** [comment]
感谢产品组，这个问题已经修复，已经测试过，1、leader节点直接掉线，然后恢复；2、leader节点、flower节点直接掉线，正常关闭，再启动，均能保持节点数据完整；3、在主节点恢复之前单节点备份过来的数据，均能同步到所有节点。目前达到一个高可用的状态；

**Comment by imbajin** [comment]
感谢反馈, 也欢迎继续测试使用, 看看其他的 case 和生产环境的稳定性等. 后续有问题也可随时反馈

**Comment by liweidong1216** [comment]
因为目前只有leader节点对元数据可以执行DELETE操作，我这边使用Hugegraph-Client操作Hugegraph，有没有现成的API能够获取到Leader节点

**Comment by imbajin** [comment]
Client 这边可能还没有吧? @javeme

**Comment by liweidong1216** [comment]
期待Client能够自动获取Server的Leader节点

**Comment by javeme** [comment]
目前仅提供 REST API 及 raft-tools.sh 工具，还有贡献client代码。
 
```sh
hugegraph-0.12.0/bin/raft-tools.sh 
usage: raft-tools.sh [options]
options: 
  -l,--list-peers ${graph} ${group}                    list all peers' endpoints for graph, can be used on leader or follower node
  -g,--get-leader ${graph} ${group}                    get the leader endpoint for graph, can be used on leader or follower node
  -s,--set-leader ${graph} ${group} ${endpoint}        set the leader endpoint for graph, can be used on leader or follower node
  -t,--transfer-leader ${graph} ${group} ${endpoint}   transfer leader to specified endpoint for graph, can be used on leader node
  -a,--add-peer ${graph} ${group} ${endpoint}          add peer for graph, can be used on leader node
  -r,--remove-peer ${graph} ${group} ${endpoint}       remove peer for graph, can be used on leader node
  -h,--help                                            display help information
```
获取leader示例：
`bin/raft-tools.sh --get-leader hugegraph`

**Comment by liweidong1216** [comment]
我先测试，感谢

**Comment by javeme** [comment]
> 感谢产品组，这个问题已经修复，已经测试过，1、leader节点直接掉线，然后恢复；2、leader节点、flower节点直接掉线，正常关闭，再启动，均能保持节点数据完整；3、在主节点恢复之前单节点备份过来的数据，均能同步到所有节点。目前达到一个高可用的状态；

Close since fixed by v0.12.0-b1
